File: server/modules/remision.js
const _EmptySpace = process.vars.espacioVacio;
const _RellenoNombre = process.vars.espacioRelleno;

module.exports = {
    //Metodo para separar y las ordenes de compra para el intercambio de documentos
    formarTxt: (pedidoData) => {
        console.log("Este es mi pedidoData formar txt ", pedidoData)
        return new Promise((resolve, reject) => {
            let clienteMs = (pedidoData.clienteMs) ? pedidoData.clienteMs : undefined;
            let pedido = pedidoData.encabezado.folio;
            let cliente = pedidoData.encabezado.cliente;
            let tipodeCambio = pedidoData.encabezado.tipodecambio;
            let almacen = pedidoData.encabezado.almacen;
            let plazo = pedidoData.encabezado.plazo;
            let productos = pedidoData.productos;
            let fecha = Math.floor((Math.random() * Date.now()) + 1);
            let rutaIn, rutaOut, filename;

            if (Object.keys(process.vars.rutas).indexOf(clienteMs) >= 0) {
                rutaIn = process.vars.rutas[clienteMs].remision.in
                rutaOut = process.vars.rutas[clienteMs].remision.out
                rutaError = process.vars.rutas[clienteMs].remision.error
                filename = process.vars.rutas[clienteMs].remision.prefijo + almacen + '_' + cliente + fecha;

                if (pedidoData.facturaProveedor) {
                    rutaIn = process.vars.rutas[clienteMs].facturaProveedor.in
                    rutaOut = process.vars.rutas[clienteMs].facturaProveedor.out
                    rutaError = process.vars.rutas[clienteMs].facturaProveedor.error
                    filename = process.vars.rutas[clienteMs].facturaProveedor.prefijo + '_' + process.vars.rutas[clienteMs].factura.prefijo + almacen + '_' + cliente + fecha;
                }

            } else {
                rutaIn = process.vars.rutas.CT.remision.in
                rutaOut = process.vars.rutas.CT.remision.out
                rutaError = process.vars.rutas.CT.remision.error
                filename = process.vars.rutas.CT.remision.prefijo + almacen + '_' + cliente + fecha;
            }

            let grupos = {};
            for (let i = 0; i < productos.length; i++) {
                //AGRUPAMOS POR EL ID EMPRESA DEL PROVEEDOR DE CADA PRODUCTO O TODO EN IDCT
                console.log("Empresa ", productos[i].empresa)
              //  console.log("idCT ",process.vars.idCT)
                let nombreGrupo = productos[i].empresa ? productos[i].empresa: null;

                if(nombreGrupo == null)
                {
                    return reject("Error: Empresa invalida", nombreGrupo)
                }else
                {
                    if (!grupos[nombreGrupo] && nombreGrupo) {
                        grupos[nombreGrupo] = [];
                    }
                    if (nombreGrupo) {
                        grupos[nombreGrupo].push(
                            {
                                clave: productos[i].clave,
                                cantidad: productos[i].cantidad,
                                costo: productos[i].precioFinal,
                                moneda: productos[i].moneda
                            });
                    }
                }
            }

            let ordenes = [];
            let ordenTXT;

            for (let nombreGrupo in grupos) {
                // Cabecera del archivo para la Remision u Orden de Compra
                console.log("Este es mi pedidoData con todos los datos  ", pedidoData)

                if (pedidoData.facturaProveedor) {

                    ordenTXT = _EmptySpace.substr(0, 21) +
                        (nombreGrupo + _EmptySpace).substring(0, 15) + " " +
                        (almacen + _EmptySpace).substring(0, 3) + " " + plazo + "  4            " + pedidoData.remision + "@\r\n";

                    grupos[nombreGrupo].forEach((art) => {
                        let tc = (art.moneda == "USD") ? tipodeCambio : " 1.00";
                        ordenTXT += (art.cantidad + _EmptySpace).substring(0, 10) + " " +
                            (art.clave + _EmptySpace).substring(0, 15) + "  " +
                            (art.costo + _EmptySpace).substring(0, 15) + "16.00  0.00  " + (tc + _EmptySpace).substring(0, 5) + "@\r\n";
                    });
                } else {
                    ordenTXT = _EmptySpace.substr(0, 21) +
                        (nombreGrupo + _EmptySpace).substring(0, 15) + " " +
                        (almacen + _EmptySpace).substring(0, 3) + " " + plazo + "  2                    @\r\n";

                    grupos[nombreGrupo].forEach((art) => {
                        let tc = (art.moneda == "USD") ? tipodeCambio : " 1.00";
                        ordenTXT += (art.cantidad + _EmptySpace).substring(0, 10) + " " +
                            (art.clave + _EmptySpace).substring(0, 15) + "  " +
                            (art.costo + _EmptySpace).substring(0, 15) + "16.00  0.00  " + (tc + _EmptySpace).substring(0, 5) + "@\r\n";
                    });
                }

                //Rellenamos el nombre para conservar una longitud estandar en el nombre de l archivo
                filenameComp = (filename + '_' + nombreGrupo +'_'+pedido+  _RellenoNombre).substring(0, 49)

                ordenes.push(
                    {
                        folioPedido: pedido,
                        rutaIn: rutaIn,
                        rutaOut: rutaOut,
                        rutaError: rutaError,
                        filename: filenameComp,
                        txt: ordenTXT
                    });
            }
            return resolve(ordenes)
        });
    },

    //Ordenamos la respuesta del proceso de remision 
    ordenarRemision: (objDatos) => {    
        let remisiones = []
        let idEmpresa; 
        let filename;
        console.log("Este es mi objDatos para ordenarRemision ", JSON.stringify(objDatos))    
        objDatos.forEach(remision => {
            
            let error = remision ? remision.error: null

            console.log("Este es mi error ", error)
            if(error)
            {

            }else
            {
                if(remision.file.Documento.Compra && remision.file.Documento.Compra[0].Tipo[0] == 'R')
            {
                    idEmpresa = remision.filename.split('_');
                    idEmpresa = idEmpresa[2];
                    filename = remision.filename
                    remisiones.push({ idEmpresa: idEmpresa ? idEmpresa: null, remision: remision.file ? (remision.file.Documento ? (remision.file.Documento.Compra ? (remision.file.Documento.Compra[0] ? (remision.file.Documento.Compra[0].Folio ? (remision.file.Documento.Compra[0].Folio[0] ? remision.file.Documento.Compra[0].Folio[0]: null):null):null):null):null):null, filename: filename })
            }
            }
            
        })
        
        objDatos = objDatos[0]
         objDatos = {
            folioPedido: objDatos.folioPedido,
            remisiones: remisiones,
            error: (objDatos.error) ? true : null,
            errorMensaje: (objDatos.errorMensaje) ? objDatos.errorMensaje : null
        }
        
        return objDatos;
    }

};

File: server/modules/python/remision.py

#from fastapi import HTTPException
#from modules.writeTXT import *

import json 
import random 
import math
import time
import sys
import os


with open("/usr/share/nginx/html/PRIVADA/server/config/_config.json") as outfile:
    #print("Este es mi archivo ", outfile)
    rutas = json.load(outfile)
    espacioVacio = rutas["prod"]["espacioVacio"]
    espacioRelleno = rutas["prod"]["espacioRelleno"]
    

def formarTxt(pedido):
    
    #pedido_data = json.dumps(pedido)
    pedido_data = json.loads(pedido)
    
    
        
    clienteMs = pedido_data["clienteMs"] if "clienteMs" in pedido_data else None
    pedido = pedido_data["encabezado"]["folio"]
    cliente = pedido_data["encabezado"]["cliente"]
    tipodeCambio = pedido_data["encabezado"]["tipodecambio"]
    almacen = pedido_data["encabezado"]["almacen"]
    plazo = pedido_data["encabezado"]["plazo"]
    productos = pedido_data["productos"]
    fecha = math.floor(random.random() *  int(round(time.time() * 1000)))

    rutaIn = rutas["prod"]["rutas"]["CT"]["remision"]["in"]
    rutaOut = rutas["prod"]["rutas"]["CT"]["remision"]["out"]
    rutaError = rutas["prod"]["rutas"]["CT"]["remision"]["error"]
    filename = rutas["prod"]["rutas"]["CT"]["remision"]["prefijo"] + almacen + '_' + cliente + str(fecha) 

    grupos = {}
    for producto in productos:
        # AGRUPAMOS POR EL ID EMPRESA DEL PROVEEDOR DE CADA PRODUCTO O TODO EN IDCT
        
        # print("idCT ", process.vars.idCT)
        nombreGrupo = producto['empresa'] if producto['empresa'] else None
        
        if nombreGrupo is None:
            print("Error")
        else:
            if nombreGrupo and nombreGrupo not in grupos:
                grupos[nombreGrupo] = []
            if nombreGrupo:
                grupos[nombreGrupo].append({
                    'clave': producto['clave'],
                    'cantidad': producto['cantidad'],
                    'costo': producto['precioFinal'],
                    'moneda': producto['moneda']
                })

    

    ordenes = []
    ordenTXT = ''
    _EmptySpace = "                                                            "
    _RellenoNombre = "_RRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRRR"
    for nombreGrupo in grupos:
        # Cabecera del archivo para la Remision u Orden de Compra
    

        facturaProveedor = pedido_data["facturaProveedor"] if "facturaProveedor" in pedido_data else None
    


    if facturaProveedor is not None:
        ordenTXT = _EmptySpace[:21] + (nombreGrupo + _EmptySpace)[:15] + " " + (almacen + _EmptySpace)[:3] + " " + plazo + "  4            " + pedido_data['remision'] + "@\r\n"
    else:
        
        
        ordenTXT = _EmptySpace[:21] + (nombreGrupo + _EmptySpace)[:15] + " " + (almacen + _EmptySpace)[:3] + " " + str(plazo) + "  2                    @\r\n"
    #print("Este es mi ordenTxt ", ordenTXT)
    for art in grupos[nombreGrupo]:
    
        tc = tipodeCambio if art['moneda'] == "MXN" else " 1.00"
    
        
        
        ordenTXT += (str(art["cantidad"]) + _EmptySpace[:10]  + " " +
        str(art['clave']) + _EmptySpace[:5] + "  " +str(art['costo']) + _EmptySpace[:8] + "16.00  0.00  " + (str(tc) + _EmptySpace)[:5] + "@\r\n")
        
        
    # Rellenamos el nombre para conservar una longitud estandar en el nombre de l archivo
    filenameComp = (filename + '_' + nombreGrupo +'_'+pedido+  _RellenoNombre)[:49]

    ordenes.append({
        'folioPedido': pedido,
        'rutaIn': rutaIn,
        'rutaOut': rutaOut,
        'rutaError': rutaError,
        'filename': filenameComp,
        'txt': ordenTXT
    })


    print(ordenes)

    return ordenes


#print(formarTxt(sys.argv[1]))


def crearArchivos(data):    
      
  archivos_write = []
  documentos = data
  
  for archivo in documentos:
              ruta_in = archivo['rutaIn']
              filename = archivo['filename']
              txt = archivo['txt']
              print("Esta es mi ruta in ", ruta_in)
              print("Este es mi filename ", filename)
              print("Este es mi txt ", txt)
              
              ruta_archivo = ruta_in + filename + ".txt"
              
              with open(ruta_archivo, "a") as f:
                print("Este es mi f ", f)
                f.writelines(txt)
                f.close()
                if f.closed:
                    print("Entre al if ")
                else:
                    print("Error else")
                                
  archivos_write.append(archivo)
  print("Este es mi archivos write ", archivos_write)
  return archivos_write
#remisionestxt = [{'folioPedido': 'W01-33333', 'rutaIn': '/datos/webpage/Compras/in/', 'rutaOut': '/datos/webpage/Compras/out/', 'rutaError': '/datos/api_privada_produccion/server/archivos/remision/', 'filename': 'Comp01A_HMO3061998970948689_MSF001_W01-226539_RRR', 'txt': '                     MSF001          01A 00  2                    @\r\n3           ESDMSF100                 2734.67               16.00  0.00  18.91@\r\n'}]






import os
import xml.etree.ElementTree as ET
import time
from typing import List, Dict
from concurrent.futures import ThreadPoolExecutor


def read_file(filename: str, rutaIn: str) -> Dict:
    with open(os.path.join(rutaIn, filename), "r") as f:
        content = f.read()
        result = ET.parse(content)
        print("Este es mi xml ", result)
        return result


def process_file(objeto: Dict) -> Dict:
    rutaIn = objeto['rutaIn']
    rutaOut = objeto['rutaOut']
    rutaError = objeto['rutaError']
    filename = objeto['filename']
    folioPedido = objeto['folioPedido']
    encontrado = False
    contador = 0
    ruta_archivo = rutaIn + filename + '.xml'
    print("Este es mi filename", filename + '.xml')
    print("Esta es mi ruta in ", rutaIn)
    while not encontrado and contador < 150:
        try:
            with open(ruta_archivo) as outfile:
                remision = outfile.read()
                remisionFinal = ET.parse(remision)
                print("Este es mi archivo remision", remisionFinal)
                print('=============== #0 Archivo encontrado: ' + filename + '.xml')
                encontrado = True
        except Exception as e:
            contador += 1
            print(f'INTENTO {contador}')
            if contador == 150:
                mensaje = ''
                if filename.startswith('Fact'):
                    mensaje = f'NO SE PUDO PROCESAR LA FACTURA ARCHIVO: {filename}.xml'
                elif filename.startswith('Comp'):
                    mensaje = f'NO SE PUDO PROCESAR LA PROCESAR REMISION ARCHIVO: {filename}.xml'
                elif filename.startswith('PedW'):
                    mensaje = f'NO SE PUDO PROCESAR EL PEDIDO ARCHIVO: {filename}.xml'
                elif filename.startswith('Alta'):
                    mensaje = f'NO SE PUDO PROCESAR LA FACTURA PROVEEDOR ARCHIVO: {filename}.xml'
                return {
                    'folioPedido': folioPedido,
                    'error': True,
                    'errorMensaje': mensaje,
                    'rutaIn': rutaIn,
                    'rutaOut': rutaError,
                    'filename': filename + '.txt'
                }
            time.sleep(0.2)

    return {
        'folioPedido': folioPedido,
        'rutaIn': rutaIn,
        'rutaOut': rutaOut,
        'file': remisionFinal,
        'filename': filename + '.xml'
    }


def process_documents(documentos: List[Dict]) -> List[Dict]:
    encontrados = []
    with ThreadPoolExecutor() as executor:
        for objeto in documentos:
            encontrados.append(executor.submit(process_file, objeto))

    return [e.result() for e in encontrados]


# Example usage:
#documentos = [{ 'folioPedido': 'W01-226539', 'rutaIn': '/datos/webpage/Compras/in/', 'rutaOut': '/datos/webpage/Compras/out/', 'rutaError': '/datos/api_privada_produccion/server/archivos/remision/', 'filename': 'Comp01A_HMO3061527156108361_MSF001_W01-226539_RRR', 'txt': '                     MSF001          01A 00  2                    @\r\n3           ESDMSF100       2734.67        16.00  0.00  18.91@\r\n' }]
#txt = formarTxt(sys.argv[1])
#archivos = crearArchivos(txt)
documentos = [{'folioPedido': 'W01-33333', 'rutaIn': '/datos/webpage/Compras/in/', 'rutaOut': '/datos/webpage/Compras/out/', 'rutaError': '/datos/api_privada_produccion/server/archivos/remision/', 'filename': 'Comp01A_HMO3061904406154319_MSF001_W01-33333_RRRR', 'txt': '                     MSF001          01A 00  2                    @\r\n3           ESDMSF100       2734.67        16.00  0.00  18.91@\r\n'}]
result = process_documents(documentos)
print(result)
sys.stdout.flush()



       
        #return  writeFile(texto, "prueba" ,".txt", "/usr/share/nginx/html/pastrano-py/privada-py/Archivos/")
        #return {"Method": "Post", "Pedido": pedido}
    #except EmailNotValidError as e:
        # Email is not valid.
        # The exception message is human-readable.
        #print(str(e))
        #raise HTTPException(status_code=400, detail=str(e))
        #return {"Method": "Post", "Pedido": pedido, "error": str(e)}
        

